central/sensorupgrade/service/service_impl.go

```diff
@@ -66,26 +71,30 @@ func (s *service) wrapToggleResponse(config *storage.SensorUpgradeConfig) *v1.Ge
 	}
 }
 
-func defaultUpgradeConfig() *storage.SensorUpgradeConfig {
-	if getAutoUpgradeFeatureStatus() == v1.GetSensorUpgradeConfigResponse_SUPPORTED {
-		return &storage.SensorUpgradeConfig{EnableAutoUpgrade: true}
-	}
-	return &storage.SensorUpgradeConfig{EnableAutoUpgrade: false}
-}
-
-func (s *service) GetSensorUpgradeConfig(ctx context.Context, _ *v1.Empty) (*v1.GetSensorUpgradeConfigResponse, error) {
+func (s *service) getOrDefaultUpgradeConfig(ctx context.Context) (*storage.SensorUpgradeConfig, error) {
 	config, err := s.configDataStore.GetSensorUpgradeConfig(ctx)
 	if err != nil {
 		return nil, err
 	}
 	if config == nil {
 		// If there's no config in the DB, return default config according to managed central flag
-		config = defaultUpgradeConfig()
+		if getAutoUpgradeFeatureStatus() == v1.GetSensorUpgradeConfigResponse_SUPPORTED {
+			return &storage.SensorUpgradeConfig{EnableAutoUpgrade: true}, nil
+		}
+		return &storage.SensorUpgradeConfig{EnableAutoUpgrade: false}, nil
+	}
```

Comment:
The only thing that I don't like about upserting in  the table here is that the user that originated the request might not  have write access.

We can still do the same thing we did in initialize (i.e. `sac.WithAllAccess(context.Background())`) and use a permissive context, but I'm a bit paranoid about arbitrarily setting `All` access in the request path.

Reply:
Well, maybe then the upsertion should be done in `initialize()` in case the value read from db is `nil`.
We're sort of back to the original approach that existed there, but not quite. The main difference is that updating the database was done there in Singleton which isn't the most testable thing. Here we'll do that in the instance initialization which _can_ be unit tested because it actually represents the unit.
If you're adventurous, give a spike and see if you'd like the resulting code.